File: java/Pomodoro.java
/**
 * A simple command-line Pomodoro timer in Java.
 */
public class Pomodoro {
    private static final int WORK_DURATION = 25;
    private static final int SHORT_BREAK = 5;
    private static final int LONG_BREAK = 15;
    private static final int POMODOROS_BEFORE_LONG_BREAK = 4;

    private static int pomodoroCount = 0;
    private static volatile boolean isRunning = true;

    public static void main(String[] args) {
        // Setup shutdown hook for Ctrl+C
        Runtime.getRuntime().addShutdownHook(new Thread(() -> {
            if (!isRunning) {
                System.out.println("\n\nPomodoro session ended.");
                System.out.printf("Completed %d pomodoro(s). Great work!\n", pomodoroCount);
            }
        }));

        System.out.println("========================================");
        System.out.println("       POMODORO TIMER");
        System.out.println("========================================");
        System.out.printf("  Work session:  %d min\n", WORK_DURATION);
        System.out.printf("  Short break:   %d min\n", SHORT_BREAK);
        System.out.printf("  Long break:    %d min (every %d pomodoros)\n",
                LONG_BREAK, POMODOROS_BEFORE_LONG_BREAK);
        System.out.println("========================================");
        System.out.println("Press Ctrl+C to stop\n");

        try {
            while (isRunning) {
                pomodoroCount++;
                System.out.printf("\n[Pomodoro #%d]\n", pomodoroCount);

                if (!countdown(WORK_DURATION, "WORK")) break;

                waitForEnter("Press Enter to start break...");

                if (pomodoroCount % POMODOROS_BEFORE_LONG_BREAK == 0) {
                    if (!countdown(LONG_BREAK, "LONG BREAK")) break;
                } else {
                    if (!countdown(SHORT_BREAK, "SHORT BREAK")) break;
                }

                waitForEnter("Press Enter to start next pomodoro...");
            }
        } catch (InterruptedException e) {
            Thread.currentThread().interrupt();
        } finally {
            isRunning = false;
        }

        System.out.println("\n\nPomodoro session ended.");
        System.out.printf("Completed %d pomodoro(s). Great work!\n", pomodoroCount);
    }

    private static String formatTime(int seconds) {
        int mins = seconds / 60;
        int secs = seconds % 60;
        return String.format("%02d:%02d", mins, secs);
    }

    private static boolean countdown(int durationMinutes, String label) throws InterruptedException {
        int totalSeconds = durationMinutes * 60;

        System.out.printf("\n%s - %d minutes\n", label, durationMinutes);
        System.out.println("------------------------------");

        for (int remaining = totalSeconds; remaining > 0 && isRunning; remaining--) {
            System.out.printf("\r  Time remaining: %s ", formatTime(remaining));
            System.out.flush();
            Thread.sleep(1000);
        }

        System.out.print("\r  Time remaining: 00:00 \n");
        System.out.printf("\n*** %s complete! ***\u0007\n", label);
        return isRunning;
    }

    private static void waitForEnter(String message) {
        System.out.print(message);
        System.out.flush();
        try {
            System.in.read();
            // Clear remaining input
            while (System.in.available() > 0) {
                System.in.read();
            }
        } catch (Exception e) {
            // Ignore
        }
    }
}

File: php/pomodoro.php
#!/usr/bin/env php
<?php
/**
 * A simple command-line Pomodoro timer in PHP.
 */

const WORK_DURATION = 25;
const SHORT_BREAK = 5;
const LONG_BREAK = 15;
const POMODOROS_BEFORE_LONG_BREAK = 4;

$pomodoroCount = 0;
$isRunning = true;

// Setup signal handler for Ctrl+C
pcntl_signal(SIGINT, function() {
    global $pomodoroCount, $isRunning;
    $isRunning = false;
    echo "\n\nPomodoro session ended.\n";
    echo "Completed {$pomodoroCount} pomodoro(s). Great work!\n";
    exit(0);
});

function formatTime($seconds) {
    $mins = intdiv($seconds, 60);
    $secs = $seconds % 60;
    return sprintf("%02d:%02d", $mins, $secs);
}

function countdown($durationMinutes, $label) {
    global $isRunning;
    $totalSeconds = $durationMinutes * 60;

    echo "\n{$label} - {$durationMinutes} minutes\n";
    echo "------------------------------\n";

    for ($remaining = $totalSeconds; $remaining > 0 && $isRunning; $remaining--) {
        pcntl_signal_dispatch(); // Check for signals
        $timeStr = formatTime($remaining);
        echo "\r  Time remaining: {$timeStr} ";
        flush();
        sleep(1);
    }

    echo "\r  Time remaining: 00:00 \n";
    echo "\n*** {$label} complete! ***\a\n";
    return $isRunning;
}

function waitForEnter($message) {
    echo $message;
    flush();
    fgets(STDIN);
}

// Main program
echo "========================================\n";
echo "       POMODORO TIMER\n";
echo "========================================\n";
echo "  Work session:  " . WORK_DURATION . " min\n";
echo "  Short break:   " . SHORT_BREAK . " min\n";
echo "  Long break:    " . LONG_BREAK . " min (every " . POMODOROS_BEFORE_LONG_BREAK . " pomodoros)\n";
echo "========================================\n";
echo "Press Ctrl+C to stop\n\n";

while ($isRunning) {
    $pomodoroCount++;
    echo "\n[Pomodoro #{$pomodoroCount}]\n";

    if (!countdown(WORK_DURATION, "WORK")) break;

    waitForEnter("Press Enter to start break...");

    if ($pomodoroCount % POMODOROS_BEFORE_LONG_BREAK === 0) {
        if (!countdown(LONG_BREAK, "LONG BREAK")) break;
    } else {
        if (!countdown(SHORT_BREAK, "SHORT BREAK")) break;
    }

    waitForEnter("Press Enter to start next pomodoro...");
}

echo "\n\nPomodoro session ended.\n";
echo "Completed {$pomodoroCount} pomodoro(s). Great work!\n";
